src/utils.py

```diff
@@ -68,13 +84,58 @@ def analyzeFileQ2(fileName):
                 unigramDict['<S>'] = unigramDict.get('<S>', 0) + 1
                 newLine = False
         else:
+            # counting <S> and <E> for current sentence
+            unigramCount += 2
             unigramDict['<E>'] = unigramDict.get('<E>', 0) + 1
             bigramDict[lastTag + ',<E>'] = bigramDict.get(lastTag + ',<E>', 0) + 1
             lastTag = '<S>'
             newLine = True
         line = f.readline()
-    for key, value in bigramDict.items():
-        bigramDict[key] = math.log(value / float(unigramDict[key.split(',')[0]]), 10)
-    for key, value in unigramDict.items():
-        unigramDict[key] = math.log(value / float(count), 10)
-    return segmentTagsDict, unigramDict, bigramDict
+    return segmentTagsDict, unigramDict, bigramDict, unigramCount
+
+
+def viterbi(sentence, states, emissionProbabilityDict, transitionProbabilityDict):
+    v = [{}]
+    tags = []
+    for state in states:
+        transitionProb = transitionProbabilityDict['<S>'].get(state, nullProbability())
+        if emissionProbabilityDict.get(sentence[0]) != None:
+            emissionProb = emissionProbabilityDict[sentence[0]].get(state, nullProbability())
+        else:
+            emissionProb = emissionProbabilityDict.get('UNK', {}).get(state, nullProbability())
+        v[0][state] = aggregateProbabilities(transitionProb, emissionProb), ''
+    for i in range(1, len(sentence)):
+        v.append({})
+        for state1 in states:
+            maxProb = nullProbability()
+            maxState = ''
+            for state2 in states:
+                transitionProb = transitionProbabilityDict.get(state2, {}).get(state1, nullProbability())
+                if emissionProbabilityDict.get(sentence[i]) != None:
+                    emissionProb = emissionProbabilityDict[sentence[i]].get(state1, nullProbability())
+                else:
+                    emissionProb = emissionProbabilityDict.get('UNK', {}).get(state1, nullProbability())
+                currentProb = aggregateProbabilities(aggregateProbabilities(v[i - 1][state2][0], transitionProb), emissionProb)
+                if currentProb >= maxProb:
+                    maxState = state2
+                    maxProb = currentProb
+            v[i][state1] = maxProb, maxState
+    overallMaxProb = nullProbability()
+    overallMaxState = ''
+    for tag, (prob, prevTag) in v[len(v) - 1].items():
+        if prob >= overallMaxProb:
+            overallMaxProb = prob
+            overallMaxState = tag
+    if overallMaxProb == nullProbability():
+        for i in range(len(sentence)):
```

Comment:
@romantz i dont udnerstand why to put nnp in the entire sentence ?

Reply:
You compute overall probability for the sentence in a dynamic programming fashion. If at the end of the sentence you get probabilities all zero (or -infinity for logprob) that means that a valid tagging has not been found for this sentence.
I agree that we can add a heuristic where you tag the sentence up until all the probabilities are zero.